lib/Reservations.rb

```diff
@@ -0,0 +1,262 @@
+      return block
+    end
+
+    def create_block_rooms_collection
+      unless @number_of_rooms <= 5 && @number_of_rooms >= 1
+        raise ArgumentError.new("Blocks can only have between 1 and 5 rooms.")
+      end
+      rooms_available = @rooms_collection
+      @dates.each do |date|
+        list_reservations_by_date(date).each do |booking|
+          rooms_available.each do |room|
+            if room.room_number == booking.room_number
+              rooms_available.delete(room)
+            end
+          end
+        end
+      end
+      @block_rooms_collection = []
+
+      @number_of_rooms.times do |i|
+        @block_rooms_collection << rooms_available[i]
+      end
+      return @block_rooms_collection
+    end
+
+    def check_in_block(block_name, room_number)
+      @blocks_collection.each do |block|
+          if block.block_name == block_name
+            block.block_rooms_collection.each do |room|
+              if room.room_number == room_number
+              return true
+              end
+            end
+          end
+      end
+      raise ArgumentError.new("Room number #{room_number} not included in #{block_name} block.")
+      return false
+    end
+
+    def new_reservation_in_block(check_in, check_out, block_name, room_number = 0, room_rate = 200)
+      block_room_booking = Hotel::Booking.new(check_in, check_out, room_number, room_rate)
+      check_in = Date.parse(check_in)
+      check_out = Date.parse(check_out)
+
+      validate_block_dates(check_in, check_out, block_name)
+      check_in_block(block_name, room_number)
+      check_block_room_available(block_name, room_number)
+      add_block_booking_to_block(block_name, block_room_booking)
+
+      @all_reservations << block_room_booking
+      return block_room_booking
+    end
+
+    def match_block(block_name)
+      @this_block = nil
+      @blocks_collection.each do |block|
+        if block.block_name == block_name
+          @this_block = block
+        end
+      end
+      return @this_block
+    end
+
+    def validate_block_dates(check_in, check_out, block_name)
+      match_block(block_name)
+      if check_in >= @this_block.check_in && check_out <= @this_block.check_out
+        return true
+      else
+        raise ArgumentError.new("Cannot reserve for those dates - dates must be the same as block dates.")
+        return false
+      end
+    end
+
+    def add_block_booking_to_block(block_name, block_room_booking)
+      @blocks_collection.each do |block|
+        if block.block_name == block_name
+          block.booked << block_room_booking
+        end
+      end
+    end
+
+    def check_block_room_available(block_name, room_number)
+      @blocks_collection.each do |block|
+        if block.block_name == block_name
+          block.booked.each do |booking|
+            if booking.room_number == room_number
+              raise ArgumentError.new("Unable to book. Room number #{room_number} has already been booked.")
```

Comment:
This is an example of tight coupling between `Reservations` and `Block`. `Booking` reaches into the `Block` and looks at its data directly. A cleaner way to do things might be to write a method like `Block#has_room_available` and call it here.

Or you could redesign things so that when you reserve a room from a block you get the first available room (or an error if there are none). That removes a little bit of freedom from the user, but means that this check isn't needed at all.